[LTO] enable `ObjCARCContractPass` only on optimized build

DataCorrupted opened the pull request:
Summary:
\#92331 tried to make `ObjCARCContractPass` by default, but it caused a regression on O0 builds and was reverted.
This patch trys to bring that back by:

1. reverts the [revert](https://github.com/llvm/llvm-project/commit/1579e9ca9ce17364963861517fecf13b00fe4d8a).
2. `createObjCARCContractPass` only on optimized builds.

Tests are updated to refelect the changes. Specifically, all `O0` tests should not include `ObjCARCContractPass`

DataCorrupted committed:
Reapply "Run ObjCContractPass in Default Codegen Pipeline (#92331)"

This reverts commit 1579e9ca9ce17364963861517fecf13b00fe4d8a.

DataCorrupted committed:
[LTO] enable `ObjCARCContractPass` only on optimized build

Summary:
\#92331 tried to make `ObjCARCContractPass` by default, but it caused a regression on O0 builds and was reverted.
This patch trys to bring that back by:

1. reverts the [revert](https://github.com/llvm/llvm-project/commit/1579e9ca9ce17364963861517fecf13b00fe4d8a).
2. `createObjCARCContractPass` only on optimized builds.

Tests are updated to refelect the changes. Specifically, all `O0` tests should not include `ObjCARCContractPass`

Signed-off-by: Peter Rong <PeterRong@meta.com>

NuriAmari commented on llvm/include/llvm/Analysis/ObjCARCAnalysisUtils.h:
```diff
@@ -41,26 +41,26 @@ extern bool EnableARCOpts;
 /// Test if the given module looks interesting to run ARC optimization
 /// on.
 inline bool ModuleHasARC(const Module &M) {
-  return
```
Since we're not calling this anymore, I don't think we need this change.

NuriAmari commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -588,12 +588,6 @@ bool EmitAssemblyHelper::AddEmitPasses(legacy::PassManager &CodeGenPasses,
   // this also adds codegenerator level optimization passes.
   CodeGenFileType CGFT = getCodeGenFileType(Action);
 
-  // Add ObjC ARC final-cleanup optimizations. This is done as part of the
```
@cachemeifyoucan As far as I can tell, it's already the case that for non-LTO builds, this pass is only added for optimization level > 0. Is that ok to do for all builds?

I checked and directly providing clang with IR that contains `llvm.objc.clang.arc.use` and compiling it with `-O0` fails instruction selection as a result of not running this pass, so I'm assuming these intrinsics are only introduced at `-O1` or higher.

cachemeifyoucan replied:
@ahatanak probably know better. I think when the arc intrinsics are present, the pass is mandatory but I am not sure if the intrinsics are only produced for -O1 and higher.

DataCorrupted replied:
Don't run this optimization on O0 has been there for 10+ yrs since 1170b0885f635ef8566. Unfortunately I don't know the committer, but he/she/they probably did it for a reason :(

ahatanak replied:
The intrinsic calls are needed to prevent ARC optimizer from moving releases to an earlier point that can cause objects to be deallocated prematurely. clang doesn't emit the intrinsic calls if `OptimizationLevel` is zero.

DataCorrupted committed:
Revert unnecessary format changes in ObjCARCAnalysisUtils.h

Signed-off-by: Peter Rong <PeterRong@meta.com>

efriedma-quic commented:
Is there some way we can avoid the additional domtree construction? Particularly in cases where the pass doesn't do anything.

DataCorrupted commented:
> Is there some way we can avoid the additional domtree construction? Particularly in cases where the pass doesn't do anything.

This patch only changes how the pass is called. The behavior of the pass itself is probably another patch. Besides, I don't think it would be easy for `PassConfig` to get a hold of the module that we are compiling.

NuriAmari commented:
This pass is added in numerous places: https://github.com/search?q=repo%3Allvm%2Fllvm-project%20createObjCARCContractPass&type=code, and we were hoping to consolidate all the places and instead schedule it by default, or behind some kind of predicate. We haven’t been able to find a predicate that satisfies all the use cases, there always seems to be a case where the required analysis passes are scheduled when not necessarily always required.

In practice, I think this pass is already scheduled by default, or for optimized builds in almost all configurations. Auditing the places it’s added, the legacy LTO backend adds it unconditionally: https://github.com/llvm/llvm-project/blob/8364a6e108faea1a6f05f7bb53720a6fc101fed2/llvm/lib/LTO/ThinLTOCodeGenerator.cpp#L340
https://github.com/llvm/llvm-project/blob/f70f1228035c9610de38e0e376afdacb647c4ad9/llvm/lib/LTO/LTOCodeGenerator.cpp#L140 LLD for MachO injects it into the new LTO backend via a hook: https://github.com/llvm/llvm-project/blob/f70f1228035c9610de38e0e376afdacb647c4ad9/lld/MachO/LTO.cpp#L52 Clang adds it for non-LTO compiles here: https://github.com/llvm/llvm-project/blob/f70f1228035c9610de38e0e376afdacb647c4ad9/clang/lib/CodeGen/BackendUtil.cpp#L595

Granted that leaves LTO builds that are using the new LTO backend (except LLD for MachO), and distributed thin-LTO builds which currently don’t schedule it. Ultimately, our goal here is to schedule the pass for distributed thin-LTO compiles as the new test demonstrates.

We could go back to a change like: https://github.com/llvm/llvm-project/pull/92331/commits/66ddf609c0e77867ec48c17136fb80d1e482041d

That would schedule the pass unconditionally for distributed thin-LTO compiles, on top of those configurations it already runs for. The analysis passes would still be run in some cases that are not required, but I don’t see how to avoid that. If we prefer that change, over the current, we could go that direction instead.

nikic commented:
The pipeline test changes here still look problematic. Can you make the ObjCARCContract pass preserve the DT?

DataCorrupted commented:
> The pipeline test changes here still look problematic. Can you make the ObjCARCContract pass preserve the DT?

ObjCARCContractPass can't preserve the DT at static time, it may alter CFG (See more in [this patch](https://reviews.llvm.org/D92808)). Unfortunately, legacy PM won't let us preserve the DT at runtime.

Current situation:
1. ObjCARCContractPass have no overhead on [O3 builds](https://llvm-compile-time-tracker.com/compare.php?from=910292c3ac2ebe43cdbc90223c6c9702128316db&to=8cc8e5d6c6ac9bfc888f3449f7e424678deae8c2&stat=instructions%3Au).
2. On distributed thin-lto builds, the overhead is very small ([on average <0.6%](https://llvm-compile-time-tracker.com/compare.php?from=910292c3ac2ebe43cdbc90223c6c9702128316db&to=8cc8e5d6c6ac9bfc888f3449f7e424678deae8c2&stat=instructions%3Au)
3. What's more, we can’t build Darwin applications using distributed thin-lto without this change. We can gain a huge build speed improvement with distributed thin-lto. 

Do you think the compile time regressions are acceptable if this change is restricted to optimized builds, or only enabled for distributed thin-lto builds?

nikic commented:
> > The pipeline test changes here still look problematic. Can you make the ObjCARCContract pass preserve the DT?
> 
> ObjCARCContractPass can't preserve the DT at static time, it may alter CFG (See more in [this patch](https://reviews.llvm.org/D92808)). Unfortunately, legacy PM won't let us preserve the DT at runtime.

You don't need to dynamically preserve the DT, you should mark it as unconditionally preserved in AnalysisUsage. From a quick glance at the pass, the only CFG change it does is critical edge splitting, and DT is already passed while doing that, so it will be updated: https://github.com/llvm/llvm-project/blob/f1fd9d716decf3513c8ae54df1d60462eb878fb5/llvm/lib/Transforms/ObjCARC/ObjCARC.cpp#L61 You're currently just not telling the pass managers that this is the case.

DataCorrupted committed:
Add `ModuleHasARC` check back to skip the pass on non objc modules.

Signed-off-by: Peter Rong <PeterRong@meta.com>

DataCorrupted committed:
Indicate that DT is preserved in ObjCARCContractPass

Tests have been changed to show that we don't have as much DT construction as before.

DataCorrupted commented:
Interesting, I didn't know DT is preserved during critical edge splitting. Just add the line to preserve DT. Many DT Construction is removed from the tests.

nikic commented on llvm/lib/Transforms/ObjCARC/ObjCARCContract.cpp:
```diff
@@ -730,6 +740,7 @@ INITIALIZE_PASS_END(ObjCARCContractLegacyPass, "objc-arc-contract",
 void ObjCARCContractLegacyPass::getAnalysisUsage(AnalysisUsage &AU) const {
   AU.addRequired<AAResultsWrapperPass>();
   AU.addRequired<DominatorTreeWrapperPass>();
+  AU.addPreserved<DominatorTreeWrapperPass>();
```
Can you also add AAResultsWrapper and BasicAAWrapperPass here? I think that should avoid some more pipeline diffs.

DataCorrupted committed:
preserving AA analysis as well

Signed-off-by: Peter Rong <PeterRong@meta.com>

nikic reviewed (approved):
LGTM

mikaelholmen commented:
I noticed that if I compile opt with UBSan enabled on this patch and then run
```
opt -passes=objc-arc-contract bbi-98007.ll -o /dev/null
```
we get an UBSan complaint like
```
../lib/Transforms/ObjCARC/ObjCARCContract.cpp:105:39: runtime error: load of value 255, which is not a valid value for type 'bool'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../lib/Transforms/ObjCARC/ObjCARCContract.cpp:105:39 in 
```
And the bbi-98007.ll input is just
```
define void @a() {
entry:
  ret void
}
```

DataCorrupted commented:
@mikaelholmen I can't reproduce this on commit e40915b7407e with `-DLLVM_USE_SANITIZER=Undefined`. Kindly let me know how to reproduce this.

Looking from the code, I suspect that it's caused by uninitialized bool (https://github.com/llvm/llvm-project/blob/539bf499615dbbfe98deaac1021f351eaad330ea/llvm/lib/Transforms/ObjCARC/ObjCARCContract.cpp#L553)

mikaelholmen commented:
> @mikaelholmen I can't reproduce this on commit [e40915b](https://github.com/llvm/llvm-project/commit/e40915b7407eda4b370658da5c9606e310b55d19) with `-DLLVM_USE_SANITIZER=Undefined`. Kindly let me know how to reproduce this.
> 
> Looking from the code, I suspect that it's caused by uninitialized bool (
> 
> https://github.com/llvm/llvm-project/blob/539bf499615dbbfe98deaac1021f351eaad330ea/llvm/lib/Transforms/ObjCARC/ObjCARCContract.cpp#L553
> )

I think the problem is that the
```
Changed = CFGChanged = false; 
```
initialization done in
```
bool ObjCARCContract::run(Function &F, AAResults *A, DominatorTree *D) {
  if (!Run)
    return false;

  if (!EnableARCOpts)
    return false;

  Changed = CFGChanged = false;
```
is skipped if we bail out early due to "!Run" or "!EnableARCOpts".
But the pass always asks
```
  bool CFGChanged = OCAC.hasCFGChanged();
```
afterwards anyway.
